Added nanny feature and test

zhuchenwang opened the pull request.

Commit:
Added nanny feature and test

inexplicable commented on lib/index.js:
```diff
@@ -82,6 +82,7 @@ Cluster.prototype.listen = function(createApp, cb) {
             connThreshold: options.connThreshold || 10000, // recycle workers after this many connections
             uptimeThreshold: options.uptimeThreshold || 3600 * 24, // 24 hours (uptimeThreshold is in seconds)
             heartbeatInterval: options.heartbeatInterval,
+            maxHbInterval: options.maxHbInterval,
```
what about changing the name as 'maxHeartbeatDelay'

zhuchenwang replied:
Accepted

inexplicable commented on lib/process.js:
```diff
@@ -150,9 +152,12 @@ var Process = module.exports = function Process(options) {
+                    // update the last heartbeat time for the worker
+                    var workerStats = self.stats.workers[message.pid];
+                    //console.log('heartbeat ' + process.pid);
+                    workerStats.lastHbTime = Date.now();
```
prefer using full name, 'lastHeartbeatAt' maybe

zhuchenwang replied:
Accepted

inexplicable commented on lib/process.js:
```diff
@@ -194,6 +199,27 @@ var Process = module.exports = function Process(options) {
+
+                    // Check the last heartbeat time of all the workers
+                    _.each(self.stats.workers, function (workerStats, pid) {
+                        var now = Date.now();
```
why not using deathQueue function which we already have?

zhuchenwang replied:
I think deathQueue is for workers to gracefully do a suicide and a worker can still serve traffic while in the queue. However, in this case, it is very possible that the worker is blocked by some operation and not serving traffic if it does not send heartbeat information. Therefore I think in this case, we should create a replacement immediately. I looked at the code again. Maybe I can call deathWatcher here instead of self.kill(). The problem is deathWatcher is not defined in the scope where I check the heartbeat. Is it possible to move it to misc.js? Any suggestions on this?

inexplicable replied:
the concern for deathQueue is valid, and I've added some extra logic to the cluster3 branch to handle it in unity.

https://github.com/cubejs/cluster2/blob/cluster3/lib/utils.js#L335
https://github.com/cubejs/cluster2/blob/cluster3/lib/utils.js#L151

safeKill is a function which sends 'SIGINT'/'SIGTERM' to the target worker, as soon as an error of 'no such process' is detected, we emit the 'exit' event on behalf of the runaway worker.

as you can see from the deathQueue updated, it again could handle either the normal 'suicide' or the 'runaway' using timeout control.

inexplicable commented:
excellent response time, could u take a look at the comment, and it could be more beneficial if we move the 'demo' test to the nodeunit test case. thanks.

zhuchenwang committed:
Merge pull request #2 from cubejs/master

Pull the latest update

Commit:
Used deathQueue to kill the unresponsive worker and create a new successor
Changed the name convention
Added test case and modified test/lib/server.js to support the test case

Commit:
Added test case to verify the number of workers

inexplicable commented on test/nanny-feature-test.js:
```diff
@@ -0,0 +1,70 @@
+'use strict';
```
we need to convert this test case to nodeunit for now

zhuchenwang replied:
Updated the test case
https://github.com/zhuchenwang/cluster2/commit/98f4437c552810df090cbdb3fd012abecfd18449

Commit:
Converted the test case to nodeunit test case
